### PR TITLE
[1852] Add provider publish logic

### PR DESCRIPTION
**Pull request by faissalMT**
### Context
As it stands the provider publish endpoints are just stubs.  
  
### Changes proposed in this pull request
 - Publish the enrichments for a provider    
 - Sync SearchAndCompare  
  
### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by faissalMT** [comment]
[Provider #publish logic](https://trello.com/c/rFNJuzpw/1852-provider-publish-logic)

File: app/controllers/api/v2/providers_controller.rb [hunk 1]
```diff
@@ -41,7 +41,15 @@ def publish
         authorize @provider, :publish?
 
         if @provider.publishable?
-          head :ok
+          @provider.publish_enrichment(@current_user)
+
+          if has_synced?
+            head :ok
+          else
+            raise RuntimeError.new(
```
Comment (defong) [inline app/controllers/api/v2/providers_controller.rb]: can we have this instead
``` ruby
         raise RuntimeError.new(
            "#{@provider} failed to sync these courses #{@provider.syncable_courses.pluck(:course_code)}"
          )```

File: app/controllers/api/v2/providers_controller.rb [hunk 2]
```diff
@@ -107,6 +115,10 @@ def enrichment_params
             :region_code
           )
       end
+
+      def has_synced?
+        SearchAndCompareAPIService::Request.sync(@provider.courses)
```
Comment (defong) [inline app/controllers/api/v2/providers_controller.rb]: Don't think that will work...

One of those courses may not be in a ready state

File: spec/requests/api/v2/providers/publish_spec.rb [hunk 3]
```diff
@@ -36,16 +36,45 @@
...
     context 'unpublished provider with draft enrichment' do
       let(:enrichment) { build(:provider_enrichment, :initial_draft) }
-      let(:provider) {
+      let!(:provider) do
         create(
           :provider,
           organisations: [organisation],
-          enrichments: [enrichment]
+          enrichments: [enrichment],
+          courses: [create(:course), create(:course)]
```
Comment (defong) [inline spec/requests/api/v2/providers/publish_spec.rb]: build vs create?

Comment (faissalMT) [reply spec/requests/api/v2/providers/publish_spec.rb]: build breaks it.

Comment (defong) [reply spec/requests/api/v2/providers/publish_spec.rb]: odd enrichment uses build....

Comment (defong) [reply spec/requests/api/v2/providers/publish_spec.rb]: is the course created assigns the current provider or a new provider?

Comment (faissalMT) [reply spec/requests/api/v2/providers/publish_spec.rb]: Oh hang on I thought you were talking about switching the provider to `build()`, switching the courses works fine.

File: app/controllers/api/v2/providers_controller.rb (hunk 1 above)
Comment (misaka) [reply app/controllers/api/v2/providers_controller.rb]: Discussed in person with @defong, probably not worth this change ... we have other things to improve on:

- `Provider#to_s` should include the recruitment year
- `SearchAndCompareAPIService` should allow us to inspect the result to see what status code we got back, not only rely on a boolean

Comment (defong) [reply app/controllers/api/v2/providers_controller.rb]: https://trello.com/c/usEjgTt1/1877-add-recruitment-year-and-provider-code-to-coursetos

https://trello.com/c/kVWN5rBS/1876-add-recruitment-year-to-providertos

**Pull request merged**